### PR TITLE
Add VK_KHR_pipeline_binary

### PR DESCRIPTION
**Pull request by stu-s**

### CODE REVIEW

File: layers/core_checks/cc_device.cpp [hunk 1]
```diff
@@ -942,3 +956,73 @@ bool CoreChecks::PreCallValidateGetDeviceFaultInfoEXT(VkDevice device, VkDeviceF
...
+        auto pipeline = pCreateInfo->pipeline;
+        auto pipeline_state = Get<vvl::Pipeline>(pipeline);
+
+        ASSERT_AND_RETURN_SKIP(pipeline_state);
+
+        if (!(pipeline_state->create_flags & VK_PIPELINE_CREATE_2_CAPTURE_DATA_BIT_KHR)) {
+            skip |= LogError("VUID-VkPipelineBinaryCreateInfoKHR-pipeline-09607", pipeline,
+                             error_obj.location.dot(Field::pCreateInfo).dot(Field::pipeline),
+                             "called on a pipeline created without the "
+                             "VK_PIPELINE_CREATE_2_CAPTURE_DATA_BIT_KHR flag set.");
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_device.cpp:980]: I just hit this same error for another extension being worked on, I would add the following as it is easy to go

```
pipeline_info.flags |= VK_PIPELINE_CREATE_2_CAPTURE_DATA_BIT_KHR;
```

and have no compiler error and go "wait, wtf, I AM setting this flag!!!"

```suggestion
                             "VK_PIPELINE_CREATE_2_CAPTURE_DATA_BIT_KHR flag set. (Make sure you set it with VkPipelineCreateFlags2CreateInfoKHR)");
```

File: layers/core_checks/cc_device.cpp [hunk 2]
```diff
@@ -942,3 +956,73 @@ bool CoreChecks::PreCallValidateGetDeviceFaultInfoEXT(VkDevice device, VkDeviceF
...
+            skip |= LogError("VUID-VkPipelineBinaryCreateInfoKHR-device-09610", device,
+                             error_obj.location.dot(Field::pCreateInfo).dot(Field::pPipelineCreateInfo),
+                             "called when disableInternalCache is true, and pPipelineCreateInfo not nullptr.");
+        }
+
+        const auto *binary_info = vku::FindStructInPNextChain<VkPipelineBinaryInfoKHR>(pCreateInfo->pPipelineCreateInfo);
+        if (binary_info && (binary_info->binaryCount > 0)) {
+            skip |= LogError("VUID-VkPipelineBinaryCreateInfoKHR-pPipelineCreateInfo-09606", device,
+                             error_obj.location.dot(Field::pCreateInfo).dot(Field::pPipelineCreateInfo),
+                             "pPipelineCreateInfo sets binaryCount to a value greater than zero");
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_device.cpp:1011]:
```suggestion
error_obj.location.dot(Field::pCreateInfo).dot(Field::pPipelineCreateInfo).dot(Field::binaryCount), "(%" PRIu32 ") is not zero", binary_info->binaryCount);
```

File: layers/core_checks/cc_device.cpp [hunk 3]
```diff
@@ -942,3 +956,73 @@ bool CoreChecks::PreCallValidateGetDeviceFaultInfoEXT(VkDevice device, VkDeviceF
...
+
+bool CoreChecks::PreCallValidateCreatePipelineBinariesKHR(VkDevice device, const VkPipelineBinaryCreateInfoKHR *pCreateInfo,
+                                                          const VkAllocationCallbacks *pAllocator,
+                                                          VkPipelineBinaryHandlesInfoKHR *pBinaries,
+                                                          const ErrorObject &error_obj) const {
+
+    bool skip = false;
+
+    uint32_t pointerCount = 0;
+
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_device.cpp:968]: feel free to do 

```c++
const Location create_info_loc =  error_obj.location.dot(Field::pCreateInfo);
```

here, it is not that expensive and really keeps the error messages a bit simpler to read

File: layers/core_checks/cc_device.cpp [hunk 4]
```diff
@@ -942,3 +956,73 @@ bool CoreChecks::PreCallValidateGetDeviceFaultInfoEXT(VkDevice device, VkDeviceF
...
+        pointerCount++;
+    }
+
+    if (pCreateInfo->pPipelineCreateInfo != nullptr) {
+        auto *props = &phys_dev_ext_props.pipeline_binary_props;
+
+        if (!props->pipelineBinaryInternalCache) {
+            skip |= LogError("VUID-VkPipelineBinaryCreateInfoKHR-pipelineBinaryInternalCache-09609", device,
+                             error_obj.location.dot(Field::pCreateInfo).dot(Field::pPipelineCreateInfo),
+                             "called when pipelineBinaryInternalCache is false, and pPipelineCreateInfo not nullptr.");
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_device.cpp:998]: `error_obj.location.dot(Field::pCreateInfo).dot(Field::pPipelineCreateInfo),` will print out the location, so the error message should be in that order

```suggestion
                             "is not NULL, but pipelineBinaryInternalCache is false.");
```

File: layers/core_checks/cc_device.cpp [hunk 5]
```diff
@@ -942,3 +956,73 @@ bool CoreChecks::PreCallValidateGetDeviceFaultInfoEXT(VkDevice device, VkDeviceF
...
+        }
+
+        if (props->pipelineBinaryInternalCacheControl && disable_internal_pipeline_cache) {
+            skip |= LogError("VUID-VkPipelineBinaryCreateInfoKHR-device-09610", device,
+                             error_obj.location.dot(Field::pCreateInfo).dot(Field::pPipelineCreateInfo),
+                             "called when disableInternalCache is true, and pPipelineCreateInfo not nullptr.");
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_device.cpp:1004]:
```suggestion
                             "is not NULL, but disableInternalCache is true.");
```

File: layers/core_checks/cc_pipeline.cpp [hunk 6]
```diff
@@ -705,3 +712,67 @@ bool CoreChecks::ValidatePipelineShaderStage(const vvl::Pipeline &pipeline,
...
+                                                               const ErrorObject &error_obj) const {
+    auto pipeline_state = Get<vvl::Pipeline>(pInfo->pipeline);
+    bool skip = false;
+
+    ASSERT_AND_RETURN_SKIP(pipeline_state);
+
+    if (!(pipeline_state->create_flags & VK_PIPELINE_CREATE_2_CAPTURE_DATA_BIT_KHR)) {
+        skip |= LogError("VUID-VkReleaseCapturedPipelineDataInfoKHR-pipeline-09613", pInfo->pipeline, error_obj.location.dot(Field::pInfo).dot(Field::pipeline),
+                         "called on a pipeline created without the "
+                         "VK_PIPELINE_CREATE_2_CAPTURE_DATA_BIT_KHR flag set.");
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_pipeline.cpp:756]:
```suggestion
                         "VK_PIPELINE_CREATE_2_CAPTURE_DATA_BIT_KHR flag set. (Make sure you set it with VkPipelineCreateFlags2CreateInfoKHR)");
```

File: scripts/generators/object_tracker_generator.py [hunk 7]
```diff
@@ -138,6 +139,7 @@ def __init__(self,
             'vkGetSwapchainImagesKHR',
             # For tracking lifetime of linked GPL libraries
             'vkCreateGraphicsPipelines',
+			'vkCreatePipelineBinariesKHR',
```
Comment (spencer-lunarg) [inline scripts/generators/object_tracker_generator.py:142]:
```suggestion
            'vkCreatePipelineBinariesKHR',
```

**Comment by stu-s** [comment]
@spencer-lunarg any thoughts on what's causing the Max Profile CI NegativePipeline.PipelineBinaryDestroy test to fail?  The test passes on a real driver and other tests seem to call vkDestroyPipelineBinary successfully.  I'm not sure if overriding the binary create call in test_icd.cpp may be causing issues (some kind of override is needed as the autogen'd function does not compile).

**Comment by spencer-lunarg** [comment]
> any thoughts on what's causing the Max Profile CI NegativePipeline.PipelineBinaryDestroy test to fail?

I will take a look today

File: tests/unit/pipeline.cpp [hunk 8]
```diff
@@ -3654,4 +3654,890 @@ TEST_F(NegativePipeline, NoRasterizationStateDynamicRendering) {
...
+
+TEST_F(NegativePipeline, PipelineBinaryCreation5) {
+    TEST_DESCRIPTION("Test creating pipeline binaries with invalid parameters");
+    SetTargetApiVersion(VK_API_VERSION_1_1);
+    AddRequiredExtensions(VK_KHR_MAINTENANCE_5_EXTENSION_NAME);
+    AddRequiredFeature(vkt::Feature::maintenance5);
+    AddRequiredExtensions(VK_KHR_PIPELINE_BINARY_EXTENSION_NAME);
+    AddRequiredFeature(vkt::Feature::pipelineBinaries);
+    RETURN_IF_SKIP(Init());
+
```
Comment (spencer-lunarg) [inline tests/unit/pipeline.cpp:4438]: Was playing around with some properties, you will need to safe guard some of these tests (which is why I am super happy we broke them up to smaller tests!)

```c++
    VkPhysicalDevicePipelineBinaryPropertiesKHR pipeline_binary_properties = vku::InitStructHelper();
    GetPhysicalDeviceProperties2(pipeline_binary_properties);
    if (!pipeline_binary_properties.pipelineBinaryInternalCache) {
        GTEST_SKIP() << "pipelineBinaryInternalCache is VK_FALSE";
    }
```

File: tests/unit/pipeline.cpp [hunk 9]
```diff
@@ -3654,4 +3654,890 @@ TEST_F(NegativePipeline, NoRasterizationStateDynamicRendering) {
...
+    // test 0
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkPipelineBinaryCreateInfoKHR-pKeysAndDataInfo-09619");
+    vk::CreatePipelineBinariesKHR(device(), &binary_create_info, nullptr, &handlesInfo);
+    m_errorMonitor->VerifyFound();
+
+    // test > 0
+    binary_create_info.pipeline = test_pipeline4;
+    binary_create_info.pPipelineCreateInfo = &pipeline_create_info;
+
```
Comment (spencer-lunarg) [inline tests/unit/pipeline.cpp:4422]:
```suggestion
    m_errorMonitor->SetUnexpectedError("VUID-VkPipelineBinaryCreateInfoKHR-pipelineBinaryInternalCache-09609");
```

File: tests/unit/pipeline_positive.cpp [hunk 10]
```diff
@@ -1828,3 +1828,160 @@ TEST_F(PositivePipeline, PipelineMissingFeaturesDynamic) {
...
+
+        VkPipelineBinaryHandlesInfoKHR handlesInfo = vku::InitStructHelper();
+        handlesInfo.pPipelineBinaries = nullptr;
+
+        err = vk::CreatePipelineBinariesKHR(device(), &binary_create_info, nullptr, &handlesInfo);
+        ASSERT_EQ(VK_SUCCESS, err);
+
+        if (handlesInfo.pipelineBinaryCount != 1)
+        {
+            GTEST_SKIP() << "Test doesn't support multiple binaries";
```
Comment (spencer-lunarg) [inline tests/unit/pipeline_positive.cpp:1924]:
```suggestion
            vk::DestroyPipeline(device(), test_pipeline1, nullptr);
            GTEST_SKIP() << "Test doesn't support multiple binaries";
```

Comment (stu-s) [reply tests/unit/pipeline_positive.cpp:1924]: I'll delete the binaries too

File: tests/unit/pipeline.cpp [hunk 11]
```diff
@@ -3654,4 +3654,890 @@ TEST_F(NegativePipeline, NoRasterizationStateDynamicRendering) {
...
+    VkPipelineBinaryInfoKHR binary_info = vku::InitStructHelper();
+    binary_info.binaryCount = binaries.size();
+    binary_info.pPipelineBinaries = binaries.data();
+    flags2.pNext = &binary_info;
+
+    handlesInfo.pPipelineBinaries = nullptr;
+    binary_create_info.pipeline = VK_NULL_HANDLE;
+    binary_create_info.pPipelineCreateInfo = &pipeline_create_info;
+
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-VkPipelineBinaryCreateInfoKHR-pPipelineCreateInfo-09606");
```
Comment (spencer-lunarg) [inline tests/unit/pipeline.cpp:4497]: this doesn't work, `binary_create_info->pNext` is null

Comment (stu-s) [reply tests/unit/pipeline.cpp:4497]: binary_create_info->pNext must be NULL (VUID-VkPipelineBinaryCreateInfoKHR-pNext-pNext)

Comment (stu-s) [reply tests/unit/pipeline.cpp:4497]: Strange.  I'm seeing one on our implementation, and I think that the ASSERT on the previous CreatePipelineBinary call would have gone off for 0.

Comment (stu-s) [reply tests/unit/pipeline.cpp:4497]: Ah, I'm wondering if the implementation of CreatePipelineBinariesKHR in test_icd is completely wrong...

Comment (spencer-lunarg) [reply tests/unit/pipeline.cpp:4497]: I think it is fine, just needs a null check as mentioned here https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/8447#discussion_r1732959747

Comment (stu-s) [reply tests/unit/pipeline.cpp:4497]: I think that it needs to return a non-zero pipelineBinaryCount too.

Comment (stu-s) [reply tests/unit/pipeline.cpp:4497]: And GetPipelineBinaryDataKHR probably needs something too :|

**Review by spencer-lunarg — APPROVED** [review]
@stu-s good to go once you rebase on https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/8448 to make sure

**Comment by stu-s** [comment]
> @stu-s good to go once you rebase on #8448 to make sure

Done.  Thanks for all the help!

**Pull request merged**